### PR TITLE
Modified subscribe() to validate topics before sending subscriptions to MQTT servers

### PR DESCRIPTION
**Pull request by anhldbk**
This is the PR addressing issue #172

### CODE REVIEW

File: lib/client.js [hunk 1]
```diff
@@ -403,6 +449,12 @@ MqttClient.prototype.subscribe = function () {
     callback = nop;
   }
 
+  invalidTopic = validateTopics(obj);
+  if ( null !== invalidTopic ) {
+    callback('Invalid topic ' + invalidTopic);
```
Comment (mcollina) [inline lib/client.js]: Use a full error here, not a string.

File: lib/client.js [hunk 2]
```diff
@@ -57,6 +57,51 @@ function storeAndSend (client, packet, cb) {
...
+ * @returns {String} If the topics is valid, returns null. Otherwise, returns the invalid one
+ */
+function validateTopics (topics) {
+  for (var i = 0; i < topics.length; i++) {
+    if ( !validateTopic(topics[i]) ) {
+      return topics[i];
+    }
+  }
+  return null;
+}
```
Comment (mcollina) [inline lib/client.js]: It would be probably good to move these into another file, `client.js` is already too big.

File: test/abstract_client.js [hunk 3]
```diff
@@ -255,6 +255,30 @@ module.exports = function (server, config) {
...
+
+    it('otherwise, should return an error (via callbacks)', function (done) {
+      var client = connect();
+      client.subscribe( ['#/event', 'event#', 'event+'], function (err) {
+        if (err) {
+          return done();
+        }
+        done(new Error('Validations do NOT work'));
+      });
+    });
```
Comment (mcollina) [inline test/abstract_client.js]: Can you please split this test into multiple cases? _all_ of those are invalid, but this test verifies that _just one_ is invalid.

File: test/abstract_client.js [hunk 4]
```diff
@@ -255,6 +255,30 @@ module.exports = function (server, config) {
     });
   });
 
+  describe('Topic validations when subscribing', function () {
+
+    it('should be ok for well-formated topics', function (done) {
+      var client = connect();
+      client.subscribe( ['+', '+/event', 'event/+', '#', 'event/#'], function (err) {
```
Comment (mcollina) [inline test/abstract_client.js]: can you add also `a/b/c/d`?

Comment (sandro-k) [inline test/abstract_client.js]: and `a/+/c` - thanks

File: lib/client.js (hunk 2 above)
Comment (anhldbk) [inline lib/client.js]: Fixed! I've moved the implementations into `validations.js`

File: lib/client.js (hunk 1 above)
Comment (anhldbk) [inline lib/client.js]: Fixed

File: test/abstract_client.js (hunk 3 above)
Comment (anhldbk) [inline test/abstract_client.js]: Tests are divided into smaller ones.

File: test/abstract_client.js (hunk 4 above)
Comment (anhldbk) [inline test/abstract_client.js]: @mcollina , @sandro-k : All tests are included in.

File: test/abstract_client.js [hunk 5]
```diff
@@ -255,6 +255,80 @@ module.exports = function (server, config) {
...
+      client.subscribe( 'system/+/#/event', function (err) {
+        if (err) {
+          return done();
+        }
+        done(new Error('Validations do NOT work'));
+      });
+    });
+
+
+
```
Comment (mcollina) [inline test/abstract_client.js]: Can you please remove the double newlines here and before?

Comment (anhldbk) [inline test/abstract_client.js]: I removed them :)

**Comment by mcollina** [comment]
This is ok for me. Any other opinion?

**Comment by anhldbk** [comment]
@mcollina So yes?

**Pull request merged**